### ISSUE TITLE
Inlining in shared generic code leaves unused temp behind

### ISSUE THREAD

**Issue by jkotas** [dotnet/coreclr#9547]
Forked from https://github.com/dotnet/coreclr/pull/9541#issuecomment-279266455.

The code for `List<__Canon>::Clear()` after #9540:

```asm
G_M58440_IG01:
       56                   push     rsi
       4883EC30             sub      rsp, 48
       48894C2428           mov      qword ptr [rsp+28H], rcx
       488BF1               mov      rsi, rcx

G_M58440_IG02:
       488B0E               mov      rcx, qword ptr [rsi]        ;** ?? --- redundant load?        
       448B4618             mov      r8d, dword ptr [rsi+24]     ; int size = _size;
       33C9                 xor      ecx, ecx                    ;** 0 ----^ clears earlier move
...
```

The redundant load is left over temp from optimized out dictionary lookup. Can we get rid of it?

Jit Dump: [ListClear.txt](https://github.com/dotnet/coreclr/files/769899/ListClear.txt)

**Comment by jkotas** [comment]
cc @dotnet/jit-contrib

**Comment by jkotas** [comment]
Notice that the frame setup is unnecessary as well because of the method is leaf method now. Not sure whether it is side-effect of the unused dictionary temp, or whether it is separate issue.

**Comment by mikedn** [comment]
It looks like `rsi` contains `this`. As far as the JIT is concerned `this` may be null so that load cannot be eliminated because it has an associated side effect.

**Comment by AndyAyersMS** [comment]
As for why there's a frame, it looks like the jit sets `lvaGenericsContextUsed` when it creates the initial runtime lookup and does not reconsider it later on when the use is deleted. Setting this flag triggers reporting of the context in the gc info (via a copy of `this`, kept alive in a stack slot) which in turn requires a frame.

It might be possible to recompute this flag post-optimization, or perhaps ref count the uses.

**Comment by mikedn** [comment]
>As for why there's a frame, it looks like the jit sets lvaGenericsContextUsed when it creates the initial runtime lookup and does not reconsider it later on when the use is deleted

Similar to #1691

**Comment by VSadov** [comment]
Yes. this seems a dupe of #1691

It is kind of annoying. It did push me in some cases to use instance methods instead of static ones, even though there was no use of `this`.

Like in: https://github.com/VSadov/NonBlocking/blob/master/src/NonBlocking/ConcurrentDictionary/DictionaryImpl.cs#L57

**Comment by jkotas** [comment]
Ok, let's use #1691 to track it then.
